### PR TITLE
Hari/sundials boxinteg

### PR DESCRIPTION
**Pull request by hsitaram**
Adding chemistry integration for sundials which takes in array4s instead of flattened arrays. It should give better performance on GPU for larger problem sizes. Although, i didn't see much with the regtest/PMF case. We may have to merge Jon's "managed memory removal" PR first and I can merge it into this branch.

### CODE REVIEW

**Comment by marchdf** [comment]
This looks fine, but 2 questions: 
- why is PMF CI failing?
- do we need another ifdef or can we do without? I know Jon is trying to minimize those...

**Comment by hsitaram** [comment]
I am looking at why CI is failing. Nothing wrong with CI, I am making a mistake.

**Review by jrood-nrel — COMMENTED** [review]
I have three comments:

1. Can this be a `pelec.chem_integrator=3` that is under `USE_SUNDIALS_PP` instead of a new ifdef? I can add cvode to the SUNDIALS submodule.
2. I don't think this needs to be CUDA-specific. Can you make it more general like #232 ?
3. I'm guessing something is going out of bounds since the debug build is failing.

**Comment by hsitaram** [comment]
i like the idea of the run-time option for pelec.chem_integrator. Let me first get CI to go through and I will make that change. I have merged with #232, so now using AMREX_USE_GPU

**Comment by hsitaram** [comment]
I have addressed everything - no more AMREX_USE_CUDAs, new pelec.chem_integrator option (number 3), CI should go through now

File: SourceCpp/React.cpp [hunk 1]
```diff
@@ -194,61 +234,84 @@ PeleC::react_state(
...
-              d_re_src_in[offset] = rhoedot_ext;
+                if (captured_clean_react_massfrac == 1) {
+                  clip_normalize_rYarr(i, j, k, sold_arr, rhoY);
+                }
+              }
             });
 
-          amrex::Gpu::streamSynchronize();
+#ifdef AMREX_USE_GPU
+          cuda_status = cudaStreamSynchronize(amrex::Gpu::gpuStream());
```
Comment (jrood-nrel) [inline SourceCpp/React.cpp]: This is still a cuda call.

File: SourceCpp/React.cpp [hunk 2]
```diff
@@ -146,6 +175,14 @@ PeleC::react_state(
           amrex::Real chemintg_cost;
           amrex::Real current_time = 0.0;
 
+#ifdef AMREX_USE_GPU
+          cudaError_t cuda_status = cudaSuccess;
```
Comment (jrood-nrel) [inline SourceCpp/React.cpp]: Can we remove this?

File: SourceCpp/React.cpp [hunk 3]
```diff
@@ -135,7 +162,9 @@ PeleC::react_state(
                 nsubsteps_max, nsubsteps_guess, errtol, do_update,
                 captured_clean_react_massfrac);
             });
-        } else if (chem_integrator == 2) {
+        }
+
+        else if (chem_integrator == 2 or chem_integrator == 3) {
```
Comment (jrood-nrel) [inline SourceCpp/React.cpp]: Prefer `||`

File: SourceCpp/React.cpp [hunk 4]
```diff
@@ -103,6 +128,7 @@ PeleC::react_state(
       const int do_update = react_init ? 0 : 1;
 
       const int captured_clean_react_massfrac = clean_react_massfrac;
+      const int captured_chem_integrator = chem_integrator;
```
Comment (jrood-nrel) [inline SourceCpp/React.cpp]: Can you reduce the scope of this so it doesn't cause an unused variable warning?

**Comment by jrood-nrel** [comment]
Thanks @hsitaram . I made a few more comments, but they're minor. I have one more question. It looks like the PelePhysics submodule would get changed with this merge. Is that necessary?

File: SourceCpp/React.cpp (hunk 1 above)
Comment (hsitaram) [reply SourceCpp/React.cpp]: removed this

File: SourceCpp/React.cpp (hunk 3 above)
Comment (hsitaram) [reply SourceCpp/React.cpp]: sure. it is now "||"

File: SourceCpp/React.cpp (hunk 4 above)
Comment (hsitaram) [reply SourceCpp/React.cpp]: I moved it into the sundials section

**Review by jrood-nrel — APPROVED** [review]

**Comment by marchdf** [comment]
@hariswaran I know this was merged, but it feels like we need a CI check and/or a regression test exercising this new capability. Do you think you could add that?

**Pull request merged**